--- ../../../../../ext/lib-BQ27421/bq27421.c	2024-04-28 22:24:36
+++ ./modified_sources/bq27421.c	2024-05-13 14:39:25
@@ -1,5 +1,5 @@
 #include "bq27421.h"
-#include "i2c.h"
+// #include "i2c.h"
 
 bool bq27421_i2c_command_write( uint8_t command, uint16_t data )
 {
@@ -173,8 +173,12 @@
     
 
     designEnergy_mWh = 3.7 * designCapacity_mAh;
+    if(taperCurrent_mA)
+    {
         taperRate = designCapacity_mAh / ( 0.1 * taperCurrent_mA );
+    }
 
+
     // Unseal gauge
     bq27421_i2c_control_write( BQ27421_CONTROL_UNSEAL );
     bq27421_i2c_control_write( BQ27421_CONTROL_UNSEAL );
@@ -183,11 +187,15 @@
     bq27421_i2c_control_write( BQ27421_CONTROL_SET_CFGUPDATE );
 
     // Poll flags
+    int max_delay = 2000;
     do
     {
         bq27421_i2c_command_read( BQ27421_FLAGS_LOW, &flags );
         if( !(flags & 0x0010) )
         {
+            if(max_delay <= 0)
+                return false;
+            max_delay -= 50;
             HAL_Delay( 50 );
         }
     }
@@ -231,12 +239,18 @@
     // Update design energy
     block[12] = (uint8_t)( designEnergy_mWh >> 8 );
     block[13] = (uint8_t)( designEnergy_mWh & 0x00FF );
+    if(terminateVoltage_mV)
+    {
         // Update terminate voltage
         block[16] = (uint8_t)( terminateVoltage_mV >> 8 );
         block[17] = (uint8_t)( terminateVoltage_mV & 0x00FF );
+    }
+    if(taperCurrent_mA)
+    {
         // Update taper rate
         block[27] = (uint8_t)( taperRate >> 8 );
         block[28] = (uint8_t)( taperRate & 0x00FF );
+    }
 
     // Calculate new checksum
     checksumNew = 0x00;
@@ -359,11 +373,15 @@
     bq27421_i2c_control_write( BQ27421_CONTROL_SOFT_RESET );
 
     // Poll flags
+    max_delay = 2000;
     do
     {
         bq27421_i2c_command_read( BQ27421_FLAGS_LOW, &flags );
         if( !(flags & 0x0010) )
         {
+            if(max_delay <= 0)
+                return false;
+            max_delay -= 50;
             HAL_Delay( 50 );
         }
     }
