--- ../../../../../ext/LoRaMac-node/src/mac/LoRaMacClassB.c	2024-10-01 14:57:12
+++ ./modified_sources/LoRaMacClassB.c	2024-10-31 16:31:27
@@ -31,6 +31,8 @@
 
 #ifdef LORAMAC_CLASSB_ENABLED
 
+void lm_radio_process_lock(void);
+void lm_radio_process_unlock(void);
 
 /*
  * LoRaMac Class B Context structure
@@ -1068,6 +1070,7 @@
                 pingSlotRxConfig.RxSlot = RX_SLOT_WIN_CLASS_B_PING_SLOT;
                 pingSlotRxConfig.NetworkActivation = *Ctx.LoRaMacClassBParams.NetworkActivation;
 
+                lm_radio_process_lock();
                 RegionRxConfig( *Ctx.LoRaMacClassBParams.LoRaMacRegion, &pingSlotRxConfig, ( int8_t* )&Ctx.LoRaMacClassBParams.McpsIndication->RxDatarate );
 
                 if( pingSlotRxConfig.RxContinuous == false )
@@ -1078,6 +1081,7 @@
                 {
                     Radio.Rx( 0 ); // Continuous mode
                 }
+                lm_radio_process_unlock();
             }
             else
             {
@@ -1250,6 +1254,7 @@
                 multicastSlotRxConfig.RxSlot = RX_SLOT_WIN_CLASS_B_MULTICAST_SLOT;
                 multicastSlotRxConfig.NetworkActivation = *Ctx.LoRaMacClassBParams.NetworkActivation;
 
+                lm_radio_process_lock();
                 RegionRxConfig( *Ctx.LoRaMacClassBParams.LoRaMacRegion, &multicastSlotRxConfig, ( int8_t* )&Ctx.LoRaMacClassBParams.McpsIndication->RxDatarate );
 
                 if( multicastSlotRxConfig.RxContinuous == false )
@@ -1260,6 +1265,7 @@
                 {
                     Radio.Rx( 0 ); // Continuous mode
                 }
+                lm_radio_process_unlock();
             }
             else
             {
