--- ../../../../ext/micropython/ports/esp32/machine_sdcard.c	2025-11-15 20:45:10.589142841 +0100
+++ ./modified_sources/machine_sdcard.c	2025-11-25 02:29:37.573917867 +0100
@@ -367,6 +367,10 @@
         DEBUG_printf("  Setting up SDMMC slot configuration");
         sdmmc_slot_config_t slot_config = SDMMC_SLOT_CONFIG_DEFAULT();
 
+        slot_config.clk = 42;
+        slot_config.cmd = 40;
+        slot_config.d0 = 41;
+
         // Stronger external pull-ups are still needed but apparently
         // it is a good idea to set the internal pull-ups anyway.
         // slot_config.flags = SDMMC_SLOT_FLAG_INTERNAL_PULLUP;
