--- ../../../../ext/micropython/ports/esp32/uart.c	2025-11-15 20:45:10.597055658 +0100
+++ ./modified_sources/uart.c	2025-11-25 02:29:37.523704790 +0100
@@ -30,6 +30,7 @@
 #include "py/runtime.h"
 #include "py/mphal.h"
 #include "uart.h"
+#include "board_hooks.h"
 
 #if MICROPY_HW_ENABLE_UART_REPL
 
@@ -82,6 +83,7 @@
     size_t remaining = len;
     uint32_t written = 0;
     // TODO add a timeout
+    __hook_mpy_uart_stdout_access_lock();
     for (;;) {
         uart_hal_write_txfifo(&repl_hal, (const void *)str, remaining, &written);
 
@@ -92,6 +94,7 @@
         str += written;
         ulTaskNotifyTake(pdFALSE, 1);
     }
+    __hook_mpy_uart_stdout_access_unlock();
     return len;
 }
 
@@ -108,12 +111,19 @@
     uart_hal_read_rxfifo(&repl_hal, rbuf, &len);
 
     for (int i = 0; i < len; i++) {
+        uint8_t c = rbuf[i];
+        #ifdef __hook_mpy_uart_irq_custom_char_handler
+        // -- hook over character handling mechanism
+        __hook_mpy_uart_irq_custom_char_handler(c);
+        #else
+        // -- retain the original micropython character handling mechanism
         if (rbuf[i] == mp_interrupt_char) {
             mp_sched_keyboard_interrupt();
         } else {
             // this is an inline function so will be in IRAM
             ringbuf_put(&stdin_ringbuf, rbuf[i]);
         }
+        #endif
     }
 }
 
